File: core/src/main/java/cz/vhromada/common/service/MovableService.java
package cz.vhromada.common.service;

import java.util.List;

import cz.vhromada.common.Movable;

/**
 * An interface represents service for movable data.
 *
 * @param <T> type of data
 * @author Vladimir Hromada
 */
public interface MovableService<T extends Movable> {

    /**
     * Creates new data.
     */
    void newData();

    /**
     * Returns list of data.
     *
     * @return list of data
     */
    List<T> getAll();

    /**
     * Returns data with ID or null if there aren't such data.
     *
     * @param id ID
     * @return data with ID or null if there aren't such data
     * @throws IllegalArgumentException if ID is null
     */
    T get(Integer id);

    /**
     * Adds data. Sets new ID and position.
     *
     * @param data data
     * @throws IllegalArgumentException if data is null
     */
    void add(T data);

    /**
     * Updates data.
     *
     * @param data new value of data
     * @throws IllegalArgumentException if data is null
     */
    void update(T data);

    /**
     * Removes data.
     *
     * @param data data
     * @throws IllegalArgumentException if data is null
     */
    void remove(T data);

    /**
     * Duplicates data.
     *
     * @param data data
     * @throws IllegalArgumentException if data is null
     */
    void duplicate(T data);

    /**
     * Moves data in list one position up.
     *
     * @param data data
     * @throws IllegalArgumentException if data is null
     */
    void moveUp(T data);

    /**
     * Moves data in list one position down.
     *
     * @param data data
     * @throws IllegalArgumentException if data is null
     */
    void moveDown(T data);

    /**
     * Updates positions.
     */
    void updatePositions();

}

File: common-core/src/main/kotlin/com/github/vhromada/common/service/MovableService.kt
package com.github.vhromada.common.service

import com.github.vhromada.common.domain.AuditEntity
import java.util.Optional

/**
 * An interface represents service for audible and movable data.
 *
 * @param <T> type of data
 * @author Vladimir Hromada
 */
interface MovableService<T : AuditEntity> {

    /**
     * Returns list of data.
     *
     * @return list of data
     */
    fun getAll(): List<T>

    /**
     * Creates new data.
     */
    fun newData()

    /**
     * Returns data with ID.
     *
     * @param id ID
     * @return data with ID
     */
    fun get(id: Int): Optional<T>

    /**
     * Adds data. Sets new ID and position.
     *
     * @param data data
     */
    fun add(data: T)

    /**
     * Updates data.
     *
     * @param data new value of data
     */
    fun update(data: T)

    /**
     * Removes data.
     *
     * @param data data
     */
    fun remove(data: T)

    /**
     * Duplicates data.
     *
     * @param data data
     */
    fun duplicate(data: T)

    /**
     * Moves data in list one position up.
     *
     * @param data data
     */
    fun moveUp(data: T)

    /**
     * Moves data in list one position down.
     *
     * @param data data
     */
    fun moveDown(data: T)

    /**
     * Updates positions.
     */
    fun updatePositions()

}
